snapshots/windows/common.go

```diff
@@ -0,0 +1,343 @@
+	}
+	return filepath.Join(s.root, "snapshots", id)
+}
+
+func (s *windowsSnapshotterBase) parentIDsToParentPaths(parentIDs []string) []string {
+	var parentLayerPaths []string
+	for _, ID := range parentIDs {
+		parentLayerPaths = append(parentLayerPaths, s.getSnapshotDir(ID))
+	}
+	return parentLayerPaths
+}
+
+// OnErrorDirectoryCleanup removes the directories if given error is nil (i.e *err == nil)
+// logs any errors if any.
+func onErrorDirectoryCleanup(ctx context.Context, err *error, dirPaths ...string) {
+	if *err != nil {
+		for _, dirPath := range dirPaths {
+			if removeErr := os.Remove(dirPath); removeErr != nil {
+				log.G(ctx).WithFields(logrus.Fields{
+					"cleanupDir":    dirPath,
+					"originalError": *err,
+					"cleanupError":  removeErr,
+				}).Warn("error while cleaning up after failure")
+			}
+		}
+	}
+}
+
+// createSnapshotCommon creates a snapshot in the metadata db with the correct snapshot info.
+// The context must be a transaction context.
+func (s *windowsSnapshotterBase) createSnapshotCommon(ctx context.Context, kind snapshots.Kind, key, parent string, opts []snapshots.Opt) (_ storage.Snapshot, _ snapshots.Info, err error) {
+	newSnapshot, err := storage.CreateSnapshot(ctx, kind, key, parent, opts...)
+	if err != nil {
+		return storage.Snapshot{}, snapshots.Info{}, errors.Wrap(err, "failed to create snapshot")
+	}
+
+	// The snapshot scratch override location could be specified in the containerd.toml or it could
+	// be specified in the container config. The one specified in the container config takes preference.
+	// Get the correct override location, update that in the snapshot snapshotInfo and save it so that all other
+	// operations will use the correct path.
+	_, snapshotInfo, _, err := storage.GetInfo(ctx, key)
+	if err != nil {
+		return storage.Snapshot{}, snapshots.Info{}, errors.Wrap(err, "failed to get snapshot info")
+	}
+
+	_, ok := snapshotInfo.Labels[labelScratchSnapshotLocation]
+	if !ok && !strings.Contains(key, snapshots.UnpackKeyPrefix) {
+		// no label provided in container config and this is a scratch snapshot
+		if s.snConfig.SnapshotterScratchLocation != "" {
+			if snapshotInfo.Labels == nil {
+				snapshotInfo.Labels = make(map[string]string)
+			}
+			snapshotInfo.Labels[labelScratchSnapshotLocation] = s.snConfig.SnapshotterScratchLocation
+			snapshotInfo, err = storage.UpdateInfo(ctx, snapshotInfo)
+			if err != nil {
+				errors.Wrap(err, "failed to write updated info")
+			}
+		}
+	}
+	return newSnapshot, snapshotInfo, nil
+}
+
+func (s *windowsSnapshotterBase) createSnapshotDirectory(ctx context.Context, snInfo snapshots.Info, snKey, snID string) (_, _ string, err error) {
+	snDir := s.getSnapshotDir(snID)
+
+	// create all parent directories first
+	if err = os.MkdirAll(filepath.Dir(snDir), 0700); err != nil {
+		return "", "", err
+	}
+
+	// Check if a different path was provided for scratch
+	snActualDir := ""
+	scratchDir, ok := snInfo.Labels[labelScratchSnapshotLocation]
```

Comment:
I guess you could also do something like:
```
snActualDir := getResolvedSnapshotDir(snID, snInfo)
if snDir != snActualDir && !strings.Contains(snKey, snapshots.UnpackKeyPrefix) {
    ...
} else {
    ...
}
return snDir, snActualDir, nil
```
This way you don't have to do the `filepath.Join(scratchDir, snID)` and Labels check, where both are done by `getResolvedSnapshotDir`.
In case when `snActual` is the same as `snDir` you'll end up returning `snDir` value instead of `""`, which isn't really an issue is it? Not sure how noisy it's going to be trying to delete the same directory twice (`onErrorDirectoryCleanup` is the only place I see where these two are passed to).

Reply:
Changed this and also updated the `onErrorDirectoryCleanup` to not log any warnings if directory doesn't exist.